arai60/word-break/step4.py

```diff
@@ -0,0 +1,39 @@
+"""
+"""
+
+# dp
+class Solution:
+    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
+        can_break = [False] * (len(s) + 1)
+        can_break[0] = True
+        max_word_len = max([len(word) for word in wordDict])
+        word_dict_set = set(wordDict)
+        for start in range(len(s) + 1):
+            if not can_break[start]:
+                continue
+            for end in range(start + 1, min(start + max_word_len + 1, len(s) + 1)):
```

Comment:
文字列の長さで for 文を回す方法と、 wordDict で for 文を回す方法があると思います。どちらが良いと思いますか？

また、 wordDict を Trie 木に入れ、 start から 1 文字ずつ辿るという方法もあると思います。ただし、こちらは常識の範囲に入っているか自信ありません。

Reply:
wordDictでfor文を回す場合、同じ長さのスライスを何度も生成する必要があり、非効率であると考えます。
よって、wordDictの最大長までのスライスを順にwordDictに入っているかどうかを調べるのが効率的だと思います。

Trie木に関して個人的に興味があるので、そちらもstep4に追加してみようかなと思います。

Reply:
D: len(wordDict), L: max_word_len とすると
O(D * L) vs O(L ^ 2) になりますかね。

Reply:
そうですね。
今回の制約だとmax_word_len<=20が保証されているので、こちらのほうが良いと考えました。

`word_lengths`みたいな配列を用意してあげて、
`end = start + word_len`のようにすると、`max_word_len`が大きい場合にも対応できそうですね

Reply:
https://docs.python.org/3/library/stdtypes.html#str.startswith
> str.startswith(prefix[, start[, end]])
> Return True if string starts with the prefix, otherwise return False. prefix can also be a tuple of prefixes to look for. With optional start, test string beginning at that position. With optional end, stop comparing string at that position.

Reply:
意図は begin があるぞです。

Reply:
失礼、start です。

Reply:
こういう機能も提供されているので、これを使うと、スライスを使う必要がなく、スライスを作るのに O(n) かける必要が無くなり、時間計算量が下がるよ。だと思います。

Reply:
内部的にはスライスを作ってるのと同じかと思ったのですが、時間計算量も下がるんでしょうか？

Reply:
strcmpでコピーは必要ないです。

Reply:
なるほど、コピーせずに前から見ていけるのが良いって感じですね
ありがとうございます🙏

Reply:
せっかくなのでwordDict + startswithの解き方も載せておきました！

Reply:
ChatGPT 先生に startswith の実装の場所を聞いたところ、 CPython の Objects\unicodeobject.c の unicode_startswith_impl() と言われました。
https://github.com/python/cpython/blob/main/Objects/unicodeobject.c#L12974
unicode_startswith_impl() の実装を読んだところ、 tailmatch() を呼び出していました。
https://github.com/python/cpython/blob/main/Objects/unicodeobject.c#L9241
tailmatch() の中では、 memcpy() でメモリの内容を比較するか、一文字ずつ比較していました。いずれにせよ、 strcmp() に近い処理が行われており、コピーは発生しないと考えて良いと思います。

Reply:
> memcpy() でメモリの内容を比較するか

-> memcmp()ですね。

Reply:
> スライスを作るのに O(n) かける必要が無くなり、時間計算量が下がるよ。

実質的に下がるでしょうが、最悪計算量は最後の文字まで運悪く一致しないだと思うので時間計算量は概ね下がらないですかね?